src/mtcp/mtcp_restart.c

```diff
@@ -1374,14 +1383,14 @@ unmap_memory_areas_and_restore_vdso(RestoreInfo *rinfo, LowerHalfInfo_t *lh_info
   /* Unmap everything except this image, vdso, vvar and vsyscall. */
   int mtcp_sys_errno;
   Area area;
-#if LINUX_VERSION_CODE > KERNEL_VERSION(3, 10, 0) 
+#if LINUX_VERSION_CODE > KERNEL_VERSION(3, 10, 0)
```

Comment:
@jungan , Shouldn't the `>`, above, be `>=`, and maybe even `>= KERNEL_VERSION(3, 11, 0)`?  At least, we wanted `>=`.

I thought that in our discussion, we saw that in Linux 3.10.x, the kernel stopped labeling vdso when we did mremap on it?  But the `#if` was testing only for `3.10.0` and not for `3.10.x`.  And I believe you were using a kernel with some version like 3.10.4, which was greater than 3.10.0, and yet the vdso label still was being lost after mremap.  And in that case, we decided to change this to:
```
#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 11, 0)
```

Reply:
Hi @gc00 , I actually found that on my CentOS 7 env, the kernel version is 3.10.0, so it will try to use `vdsoStartTmp` instead of setting it to `NULL` and relies on parsing the proc map.

The problem is, `vdsoStartTmp` becomes 0 here. The place where it started to change is here:
```
  DPRINTF("We have copied mtcp_restart to higher address.  We will now\n"
          "    jump into a copy of restorememoryareas().\n");
  rinfo.vvarStartTmp = vvarStartTmp;
  rinfo.vdsoStartTmp = vdsoStartTmp;
  DPRINTF("%p %p %p %p\n", &vvarStartTmp, vvarStartTmp, &vdsoStartTmp, vdsoStartTmp);
  /* after calling the following assembly, the address of vvarStartTmp & vdsoStartTmp changed */

#if defined(__i386__) || defined(__x86_64__)
  asm volatile (CLEAN_FOR_64_BIT(mov %0, %%esp; )
```

And it's actually because the address of the global variable `vvarStartTmp ` changed.



In terms of shall we change to `#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 11, 0)`, I don't know since I didn't try if kernel 3.10.x has the label or not.

But like @JainTwinkle also asked, if we knew we have `mremap` the vdso and we saved the new address, is it necessary to still be purely relying on parsing the proc map? Even we think it's more reliable, but it doesn't make it worse if we just init the `vdsoStart` to the address we knew we have `mremap`, we can update it if we found the `vdso` in the proc map, but I think it's fine to just use the one we saved if we didn't find it in the proc map.